### ISSUE TITLE
Fails to build with OCaml 4.12

### ISSUE THREAD

**Issue by drjdn** [camlp5/camlp5#84]
Have you tried building with 4.12?   The error output is as follows:

```
### output ###
# [...]
# the file '../boot/camlp5r' has not the right magic number: expected Caml1999X028, got Caml1999X029
# the file '../boot/camlp5r' has not the right magic number: expected Caml1999X028, got Caml1999X029
# make[3]: *** [../config/Makefile:18: diff.cmi] Error 127
# the file '../boot/camlp5r' has not the right magic number: expected Caml1999X028, got Caml1999X029
# make[3]: *** [../config/Makefile:18: ploc.cmi] Error 127
# make[3]: *** [../config/Makefile:18: extfun.cmi] Error 127
# make[3]: Leaving directory '/home/jdn/.opam/default/.opam-switch/build/camlp5.8.00.02/lib'
# make[2]: *** [Makefile:167: library] Error 2
# make[2]: Leaving directory '/home/jdn/.opam/default/.opam-switch/build/camlp5.8.00.02'
# make[1]: *** [Makefile:41: boot/camlp5] Error 2
# make[1]: Leaving directory '/home/jdn/.opam/default/.opam-switch/build/camlp5.8.00.02'
# make: *** [Makefile:159: world.opt] Error 2
```
I haven't made an attempt to triage this but there is no `4.11.2` in `camlp5/ocaml_stuff`.

Jason

**Comment by chetmurthy** [comment]
Hi.  two things:

1. the error looks like there's stuff leftover from a previous build with a different version of ocaml?  Maybe do a make clean, then configure, then make?
2. the ocaml-version-detection machinery in camlp5 should select the most-compatible "saved stuff" -- so if you build with Ocaml 4.11.2, you should get the saved stuff from ocaml 4.11.0 (or maybe 4.11.1 -- I forget).
3. there were bugs with this machinery which got fixed in newer versions of camlp5 -- can you tell me what your version of camlp5 is?  You can find it in main/pcaml.ml

**Comment by chetmurthy** [comment]
Another comment: I always test new releases of camlp5 with many versions of ocaml.  You can find the file that does it in `testsuite/tools/TRIP-TEST-ALL`.  opam also does this level of testing -- against many versions of ocaml, and on many platforms.

**Comment by chetmurthy** [comment]
Ah, you're building camlp5 8.00.02.  I am certain that passed against Ocaml 4.12.0, both on my machine, and with opam.

**Comment by drjdn** [comment]
```bash
>  camlp5 -version
8.00.02
> opam --version
2.0.9
```
I install OCaml in `/usr/local/ocaml`.  I moved `/usr/local/ocaml` to `usr/local/ocaml_bak` before installing `4.12` to `/usr/local/ocaml` via source `./configure --prefix=/usr/local/ocaml --enable-flambda`.   Then moved `.opam` to `.opam_bak` and ran:
```shell
> opam init
 [snip]
> opam install camlp5
```
and get the error above.  Should be as clean as possible and is reproducible on my machine (x86-64 running debian testing with no deb installed version of OCaml on the system).

Thanks,
Jason

**Comment by chetmurthy** [comment]
You're using opam.  So you should have created an opam switch, and be using that, right?  What is the output of "which ocaml" and "printenv OPAM_SWITCH_PREFIX" ?

what I'm saying is: "/usr/local/ocaml" shouldn't be involved at all.

**Comment by chetmurthy** [comment]
Also, the original error-message is a clear indication that there is a mixing of two different ocaml versions.  I grepped for those two magic numbers (I have opam switches for every opam-installable version from 4.02.0 thru 4.13.0~alpha02)

1. ....29 is from Ocaml 4.12.0
2. ....28 is from 4.11.1

**Comment by drjdn** [comment]
I'm not an opam expert but by default it picks up OCaml if it is on your path.  So if you have it in `/usr/local/ocaml`  you get:
```bash
> ocaml init
[snip]
<><> Creating initial switch (ocaml-system>=4.02.3) <><><><><><><><><><><><><><>

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed ocaml-system.4.12.0
∗ installed ocaml-config.2
∗ installed ocaml.4.12.0
Done.
```
which executes quickly.  If I remove the installed OCaml from path:
```bash
[snip]
<><> Creating initial switch (ocaml-base-compiler) ><><><><><><><><><><><><><><>

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ocaml-base-compiler.4.12.0] downloaded from cache at https://opam.ocaml.org/cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed ocaml-options-vanilla.1
∗ installed ocaml-base-compiler.4.12.0
∗ installed ocaml-config.2
∗ installed ocaml.4.12.0
Done.
```
and takes a bit of time as I'm assuming it is building the compiler.  If done this way `camlp5` builds correctly via `opam install camlp5`.  In the previous I get the error message above.

I understand the ast number magic and I don't think this is the issue.  I've been using opam this way with my installation of OCaml in `/usr/local/ocaml` for a few years now and `camlp5` has never failed to build.

Thanks,
Jason

**Comment by chetmurthy** [comment]
can paste the contents of `config/Makefile.cnf` ?

**Comment by chetmurthy** [comment]
> I understand the ast number magic and I don't think this is the issue.

On my machine, I have
```
chet@digg:~$ which ocaml
/usr/bin/ocaml
chet@digg:~$ ocaml -version
The OCaml toplevel, version 4.08.1
```

as well as whatever is my current opam switch, viz.
```
chet@digg:~/Hack/Camlp5/src/ALL/camlp5$ printenv | grep -i PREF
OPAM_SWITCH_PREFIX=/home/chet/Hack/Opam-2.1.0-beta4/GENERIC/4.13.0~alpha2
chet@digg:~/Hack/Camlp5/src/ALL/camlp5$ which ocaml
/home/chet/Hack/Opam-2.1.0-beta4/GENERIC/4.13.0~alpha2/bin/ocaml
chet@digg:~/Hack/Camlp5/src/ALL/camlp5$ 
```

So these are two *very* different ocaml versions, and yet, when I install camlp5 with opam, I never get these sorts of "compiler version" incompatibilities.  As you observe, your non-opam install of ocaml is in /usr/local/ocaml: I grepped below, and find no instance of that path in the sources of the current release.

```
commit 152cc58ee6a8213fbb36fbb7285d2358febb37af (HEAD -> master, tag: rel8.00.02, origin/master, origin/HEAD)
chet@digg:~/Hack/Camlp5/src/ALL/camlp5$ grep -nr /usr/local .
chet@digg:~/Hack/Camlp5/src/ALL/camlp5$ 
```

In short, I think this is an opam/ocaml environment setup problem.

**Comment by drjdn** [comment]
Sure:
```
MODE=S
EXE=
OPT=.opt
EXT_OBJ=.o
EXT_LIB=.a
OVERSION=4.12.0
VERSION=8.00.02
OCAMLC_W_Y="-w y"
WARNERR="-warn-error +A-11"
NO_PR_DIR=--no-print-directory
OLIBDIR=/usr/local/ocaml/lib/ocaml
OCLIBDIR=+compiler-libs
PREFIX=/home/jdn/.opam/default
OTOPP=+compiler-libs
OTOPW=+compiler-libs
OTOPU=$(TOP)/ocaml_stuff/4.12.0/utils
BINDIR=$(PREFIX)/bin
LIBDIR=/home/jdn/.opam/default/lib
MANDIR=/home/jdn/.opam/default/man
OCAMLN=ocaml
CAMLP5N=camlp5
```
which is in `.opam/default/.opam-switch/build/camlp5.8.00.02/config`

**Comment by drjdn** [comment]
```bash
> which ocaml
/usr/local/ocaml/bin/ocaml
> echo echo $OPAM_SWITCH_PREFIX
/home/jdn/.opam/default
```
If I do the exact same procedure with `4.11.2` or earlier `camlp5` builds fine with `opam install camlp5`.

**Comment by chetmurthy** [comment]
I have difficulty believing that this was the result of `opam install camlp5` (or any variant thereof).  The PREFIX is the dead giveaway.  Also, why is /usr/local showing up there.

How did you configure this?

**Comment by drjdn** [comment]
As above `./configure --prefix=/usr/local/ocaml --enable-flambda`

**Comment by chetmurthy** [comment]
Look: why are you resistant to just using the ocaml installation set up by opam ?  What is driving you to insist on using /usr/local/ocaml ?

**Comment by chetmurthy** [comment]
Can you see that the `Makefile.cnf` you pasted, isn't the result of a coherent environment ?

**Comment by chetmurthy** [comment]
You need to specify libdir, probably mandir also.  And the PREFIX in Makefile.cnf is not /usr/local/ocaml .

**Comment by chetmurthy** [comment]
Also: if you're building this outside of opam, why are opam environment variables set?

**Comment by drjdn** [comment]
```bash
> cat /etc/profile.d/ocaml.sh
export PATH=$PATH:/usr/local/ocaml/bin
export MANPATH=$MANPATH:/usr/local/ocaml/share/man
```
My setup is not standard and perhaps not coherent but things have worked perfectly fine until `4.12`.  Please feel free to close this issue.

**Comment by chetmurthy** [comment]
OK.  So I'm going to walk thru the logic here:

1. you want to have opam settings active in your environment
2. this will put the opam-managed switch *first* in the search paths
3. But you want to build with your own ocaml install: /usr/local/ocaml

Is this accurate?   I think it's rarely-if-ever that I build camlp5 with an ocaml installation that is *not* at the front of the search-path.  I mean, why would someone do that?  So here's what I'd suggest: -prepend- /usr/local/ocaml/bin to your PATH.  ensure that ALL relevant config-variables point at /usr/local/ocaml and NONE of them point at the opam-managed installation.

**Comment by chetmurthy** [comment]
BTW, I can understand why you might want to build a custom ocaml install.  If you *also* want to use opam, I'd suggest strongly that you can make that work: there's a way to setup an opam switch that points at a custom-built ocaml installation.  I use it when I'm debugging patches to ocaml distributions (for bugfixes).  It's documented in the "HACKING.md" file in the ocaml distribution, IIRC.

I guess what I'm saying is: if you want to use opam but also use a custom ocaml build, then try to do it in a way that is standardized by opam.  Or, y'know, don't use opam, and don't set opam set various environment variables.

**Comment by drjdn** [comment]
The only thing I'm say is that `opam` certainly picks up a locally installed version of OCaml and if you scroll up you will see `*installed ocaml-system.4.12.0*` vs. `*installed ocaml-base-compiler.4.12.0*`.  With that you can install as many `opam` packages and they install correctly in `.opam` in your home directory.  This isn't anecdotal as I've been doing this for quite some time now.  The only package that I require that is failing to build with the attempt to move to `4.12` is `camlp5` which is new and never been an issue.  I've tried this now in a `chroot jail` with a fresh install and I get the same outcome.  Anyways, don't waste your time on this as I'll figure it out.  I'm fine currently at 4.11.* and was only testing the waters and don't really care that much until the multi-core stuff comes into play.  For the record I'm not trying to bust your ass.  I look forward to your posts on the forums.  I was really excited a few years ago about a standardized AST.  The recent ppxlib stuff is OK but it hasn't lived up to the hype of the original `astlib` post!  `camlp5` is a different animal.  If this could have been sorted years ago I bet OCaml would live in a different space.  So many crappy languages... however I still have hope as the kids are bringing back modern [Fortran](https://lfortran.org) [ISO/IEC 1539:2018].  Not a joke if you actually do serious numerical work ;-).

**Comment by chetmurthy** [comment]
So, uh, I do wonder how this has been working in the past.  The camlp5 build apparatus has changed very, very little.  I'd conjecture that "quite some time now" only included 4.11.* versions of ocaml -- in that case, then there would be no version-incompatibility between the /usr/local/ocaml version, and the one managed by opam.

can you paste the output of "opam switch list" ?

**Comment by chetmurthy** [comment]
Hey, can you do that build *one* more time, and upload the entire output, along with config/Makefile.cnf ?  The entire output, not just a suffix of it.

**Comment by chetmurthy** [comment]
> I was really excited a few years ago about a standardized AST. 

Obviously I disagree with the PPX core team on how to provide compatibility support for PPX rewriters.  Basically, I don't think they have it right:
1. the OCaml AST is designed for compiler-writers, not macro-writers
2. any AST that is supposed to be compatible between OCaml versions is going to allow nonsensical instance values.
3. The thing that stays compatible from version-to-version of a language, is the surface syntax.  That's where you can achieve the most stability for PPX expander authors.

Ah, well.

**Comment by chetmurthy** [comment]
Also, looking at the opam package for "ocaml-system-4.12.0": https://github.com/ocaml/opam-repository/tree/master/packages/ocaml-system/ocaml-system.4.12.0

I don't think it does what you're thinking it does.

It might be useful if you were to write down the script you use in that chroot jail, to repro your problem.

**Comment by chetmurthy** [comment]
I know I've sounded a little irritable in appends above.  I -do- want to help you to get this working.  I just feel that .... well, you're trying to do something -very- nonstandard for opam, and also, honestly, for camlp5's build-system.

**Comment by drjdn** [comment]
No worries, I wasn't really considering that my setup was strange.  I use quite a few different languages to get things done and for all of those tool chains that are needed I build from source and put into `/usr/local`.  This isn't as crazy as it might seem as there are advantages but that is a totally different matter.  As all of these languages have started to create their own build systems/package managers, which is great, it is also difficult to keep up and do things in the 'idiomatic' way of [insert language name] ;-).  You have asked for a full dump and I will do so in the next post.

**Comment by drjdn** [comment]
This is with a fresh install of OCaml 4.12 with no .opam etc. (I think clean of any cruft that would impact the result).

```bash
> opam --version
2.0.9
> ocaml --version
The OCaml toplevel, version 4.12.0
> which ocaml
/usr/local/ocaml/bin/ocaml
```
```bash
> opam init
[NOTE] Will configure from built-in defaults.
Checking for available remotes: rsync and local, git, mercurial, darcs. Perfect!

<><> Fetching repository information ><><><><><><><><><><><><><><><><><><><><><>
[default] Initialised

<><> Required setup - please read <><><><><><><><><><><><><><><><><><><><><><><>

  In normal operation, opam only alters files within ~/.opam.

  However, to best integrate with your system, some environment variables
  should be set. If you allow it to, this initialisation step will update
  your bash configuration by adding the following line to ~/.profile:

    test -r /home/jdn/.opam/opam-init/init.sh && . /home/jdn/.opam/opam-init/init.sh > /dev/null 2> /dev/null || true

  Otherwise, every time you want to access your opam installation, you will
  need to run:

    eval $(opam env)

  You can always re-run this setup with 'opam init' later.

Do you want opam to modify ~/.profile? [N/y/f]
(default is 'no', use 'f' to choose a different file) 
A hook can be added to opam's init scripts to ensure that the shell remains in sync with the
opam environment when they are loaded. Set that up? [y/N] n

<><> Creating initial switch (ocaml-system>=4.02.3) <><><><><><><><><><><><><><>

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed ocaml-system.4.12.0
∗ installed ocaml-config.2
∗ installed ocaml.4.12.0
Done.
```
```bash
> opam install camlp5
The following actions will be performed:
  ∗ install conf-perl                   1       [required by camlp5]
  ∗ install conf-perl-string-shellquote 1       [required by camlp5]
  ∗ install conf-perl-ipc-system-simple 1       [required by camlp5]
  ∗ install camlp5                      8.00.02
===== ∗ 4 =====
Do you want to continue? [Y/n] y

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[camlp5.8.00.02] downloaded from cache at https://opam.ocaml.org/cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed conf-perl.1
∗ installed conf-perl-string-shellquote.1
∗ installed conf-perl-ipc-system-simple.1
[ERROR] The compilation of camlp5 failed at "/home/jdn/.opam/opam-init/hooks/sandbox.sh build
        make -j5 DEBUG=-g world.opt".

#=== ERROR while compiling camlp5.8.00.02 =====================================#
# context     2.0.9 | linux/x86_64 | ocaml-system.4.12.0 | https://opam.ocaml.org#5beb7871
# path        ~/.opam/default/.opam-switch/build/camlp5.8.00.02
# command     ~/.opam/opam-init/hooks/sandbox.sh build make -j5 DEBUG=-g world.opt
# exit-code   2
# env-file    ~/.opam/log/camlp5-19058-335865.env
# output-file ~/.opam/log/camlp5-19058-335865.out
### output ###
# [...]
# make[3]: *** [../config/Makefile:24: versdep.cmo] Error 127
# make[3]: *** Waiting for unfinished jobs....
# make[3]: *** [../config/Makefile:18: ploc.cmi] Error 127
# make[3]: *** [../config/Makefile:18: fstream.cmi] Error 127
# make[3]: *** [../config/Makefile:18: diff.cmi] Error 127
# make[3]: *** [../config/Makefile:18: extfun.cmi] Error 127
# make[3]: Leaving directory '/home/jdn/.opam/default/.opam-switch/build/camlp5.8.00.02/lib'
# make[2]: *** [Makefile:167: library] Error 2
# make[2]: Leaving directory '/home/jdn/.opam/default/.opam-switch/build/camlp5.8.00.02'
# make[1]: *** [Makefile:41: boot/camlp5] Error 2
# make[1]: Leaving directory '/home/jdn/.opam/default/.opam-switch/build/camlp5.8.00.02'
# make: *** [Makefile:159: world.opt] Error 2



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build camlp5 8.00.02
└─ 
┌─ The following changes have been performed
│ ∗ install conf-perl                   1
│ ∗ install conf-perl-ipc-system-simple 1
│ ∗ install conf-perl-string-shellquote 1
└─ 

The former state can be restored with:
    opam switch import
"/home/jdn/.opam/default/.opam-switch/backup/state-20210807172904.export"
```
```bash
> cat .opam/default/.opam-switch/build/camlp5.8.00.02/config/Makefile.cnf
MODE=S
EXE=
OPT=.opt
EXT_OBJ=.o
EXT_LIB=.a
OVERSION=4.12.0
VERSION=8.00.02
OCAMLC_W_Y="-w y"
WARNERR="-warn-error +A-11"
NO_PR_DIR=--no-print-directory
OLIBDIR=/usr/local/ocaml/lib/ocaml
OCLIBDIR=+compiler-libs
PREFIX=/home/jdn/.opam/default
OTOPP=+compiler-libs
OTOPW=+compiler-libs
OTOPU=$(TOP)/ocaml_stuff/4.12.0/utils
BINDIR=$(PREFIX)/bin
LIBDIR=/home/jdn/.opam/default/lib
MANDIR=/home/jdn/.opam/default/man
OCAMLN=ocaml
CAMLP5N=camlp5
```

**Comment by chetmurthy** [comment]
OK, what we need now is the build logs from opam.  Can you upload those? 

BTW, I also used to systematically build little "sandboxes" for all my various languages that I hacked on.  And different OCaml versions.  When opam came out, it took a little whlie, before I started just relying on opam for all of that.  And so the last few years, I basically never build ocaml from scratch *except* when I'm hacking on ocaml itself.

**Comment by chetmurthy** [comment]
I see you didn't do
```
    eval $(opam env)
```
that's actually important.  It might not be necessary, but typically it setups up env vars that are necessary for proper operation.

**Comment by drjdn** [comment]
I killed the `.opam` directory before reading your last reply so I don't have the exact two logs.  The reason I did this was I repeated the above with a fresh install of `OCaml 4.10.0` and it failed as well.   I was a bit surprise as I'm certain I was doing things this way when I was using the `4.10` series.  I guess I got lucky and the `ocaml-system` stuff  was lining up with the version of OCaml I was actually using at the time.  Looks like I'm going to have to put on my big boy pants and do `OCaml` stuff in the correct `opam` way.   I moved from `make` to `dune` some time ago and I have no regrets... looks like I'm going to have to get with the times.  Sorry for all the noise.
